pset/pset.go

```diff
@@ -283,6 +289,200 @@ func (p *Pset) SanityCheck() error {
+		return false, nil
+	}
+
+	return pSig.Verify(sigHash[:], pubKey), nil
+}
+
+func (p *Pset) getHashAndScriptForSignature(inputIndex int, sigHashType uint32) (
+	[]byte,
+	[]byte,
+	error,
+) {
+	var hash [32]byte
+	var script []byte
+
+	input := p.Inputs[inputIndex]
+
+	if input.NonWitnessUtxo != nil {
+		prevoutHash := p.UnsignedTx.Inputs[inputIndex].Hash
+		utxoHash := input.NonWitnessUtxo.WitnessHash()
+
+		if bytes.Compare(prevoutHash, utxoHash.CloneBytes()) == 1 {
+			return nil, nil,
+				errors.New("mon-witness utxo hash for input doesnt match the " +
+					"hash specified in the prevout")
+		}
+
+		prevoutIndex := p.UnsignedTx.Inputs[inputIndex].Index
+		prevout := input.NonWitnessUtxo.Outputs[prevoutIndex]
+		if input.RedeemScript != nil {
+			script = input.RedeemScript
+		} else {
+			script = prevout.Script
+		}
+
+		switch address.GetScriptType(script) {
+
+		case address.P2WshScript:
+			hash = p.UnsignedTx.HashForWitnessV0(
+				inputIndex,
+				input.WitnessScript,
+				prevout.Value,
+				txscript.SigHashType(sigHashType),
+			)
+			script = input.WitnessScript
+
+		case address.P2WpkhScript:
+			pay, err := payment.FromScript(
+				script,
+				&network.Liquid,
+				nil,
+			)
+			if err != nil {
+				return nil, nil, err
+			}
+			hash = p.UnsignedTx.HashForWitnessV0(
+				inputIndex,
+				pay.Script,
+				input.WitnessUtxo.Value,
+				txscript.SigHashType(sigHashType),
+			)
+		default:
+			var err error
+			hash, err = p.UnsignedTx.HashForSignature(
+				inputIndex,
+				script,
+				txscript.SigHashType(sigHashType),
+			)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+	} else if input.WitnessUtxo != nil {
+		if input.RedeemScript != nil {
+			script = input.RedeemScript
+		} else {
+			script = input.WitnessUtxo.Script
+		}
+		switch address.GetScriptType(script) {
+
+		case address.P2WpkhScript:
+			pay, err := payment.FromScript(
+				script,
+				&network.Liquid,
+				nil,
+			)
+			if err != nil {
+				return nil, nil, err
+			}
+			hash = p.UnsignedTx.HashForWitnessV0(
+				inputIndex,
+				pay.Script,
+				input.WitnessUtxo.Value,
+				txscript.SigHashType(sigHashType),
+			)
+		case address.P2WshScript:
+			hash = p.UnsignedTx.HashForWitnessV0(
+				inputIndex,
+				input.WitnessScript,
+				input.WitnessUtxo.Value,
+				txscript.SigHashType(sigHashType),
+			)
+			script = input.WitnessScript
+		default:
+			return nil, nil, errors.New("inputhas witnessUtxo but non-segwit script")
+		}
+
+	} else {
+		return nil, nil, errors.New("need a utxo input item for signing")
+	}
+
+	return hash[:], script, nil
+}
+
+//TODO: check if this is valid only for P2PKH and P2WPKH since we
+//should not be able to compare script hash with public key hash
+func (p *Pset) verifyScripForPubKey(script []byte, pubKey []byte) (bool, error) {
+	pay, err := payment.FromScript(script, &network.Liquid, nil)
+	if err != nil {
+		return false, err
+	}
+
+	pkh := payment.Hash160(pubKey)
+	res := bytes.Compare(pay.Hash, pkh)
```

Comment:
This is valid only for p2pkh/p2wpkh indeed.
In the reference implementation of [liquidJS](https://github.com/vulpemventures/liquidjs-lib/blob/master/ts_src/psbt.ts#L1124) the script is decompiled and then we check that the pkHash or the pk appears in the script. This because the script can contain:
* a full pubkey in case it's a P2PK or P2MS (in this case the script that arrives here is the redeem script containing all the pubkeys)
* a pkHash in case it's a P2PKH or a P2WPKH
* either a pkHash or a full pubkey for complex P2SH or P2WSH since for this kind of types, the script arriving here will be a redeem script (just like how happens for P2MS)


We should check if in btcutil / btcd repos there's something similar to `bScript.decompile()` that decompiles a a script into a sequence of opcodes/hashes

Reply:
I found this method [`txscript.DisasmString`](https://godoc.org/github.com/btcsuite/btcd/txscript#DisasmString) in btcd that parses a raw bytes script into a string of space separated op codes and data.
We could use this method to parse the script and then search for the pubkey of interest like:
```
scriptAsm, err = txscript.DisasmString(script)

if strings.Contains(scriptAsm, pubKey.SerializeCompressed() || strings.Contains(scriptAsm, pkHash) {
   ...
}
```

cc @tiero @sekulicd

Reply:
Done